src/porepy/models/protocol.py

```diff
@@ -0,0 +1,816 @@
+"""Contains the protocols that declare the methods and attributes that must be present
+in an instance of a PorePy model. The proper implementations of these methods can be
+found in various classes within the models directiory.
+
+"""
+
+from pathlib import Path
+from typing import TYPE_CHECKING, Callable, Literal, Optional, Protocol, Sequence
+
+import numpy as np
+import scipy.sparse as sps
+
+# Conditional importing ensures that the protocols do not mess with the runtime
+# definitions, i.e., the protocol empty method is accidentally called as a proper method
+# and returns None.
+if not TYPE_CHECKING:
+    # This branch is accessed in python runtime.
+    class PorePyModel(Protocol):
+        """This is an empty placeholder of the protocol, used mainly for type hints."""
+
+        def __init__(self, *args, **kwargs) -> None:
```

Comment:
@keileg @Yuriyzabegaev Marking this change here, since it fixes something we do not know of as of now.